xds/src/main/java/io/grpc/xds/XdsServerWrapper.java

```diff
@@ -669,48 +756,12 @@ private boolean isPermanentError(Status error) {
             new Metadata());
         return new ServerCall.Listener<ReqT>() {};
       }
-      List<ServerInterceptor> filterInterceptors = new ArrayList<>();
-      for (NamedFilterConfig namedFilterConfig : routingConfig.httpFilterConfigs()) {
-        FilterConfig filterConfig = namedFilterConfig.filterConfig;
-        Filter filter = filterRegistry.get(filterConfig.typeUrl());
-        if (filter instanceof ServerInterceptorBuilder) {
-          ServerInterceptor interceptor =
-              ((ServerInterceptorBuilder) filter).buildServerInterceptor(
-                  filterConfig, selectedOverrideConfigs.get(namedFilterConfig.name));
-          if (interceptor != null) {
-            filterInterceptors.add(interceptor);
-          }
-        } else {
-          call.close(
-              Status.UNAVAILABLE.withDescription("HttpFilterConfig(type URL: "
-                      + filterConfig.typeUrl() + ") is not supported on server-side."),
-              new Metadata());
-          return new Listener<ReqT>() {};
-        }
+      ServerInterceptor routeInterceptor = noopInterceptor;
+      Map<Route, ServerInterceptor> perRouteInterceptors = routingConfig.interceptors().get();
```

Comment:
This can "tear" with `routingConfig.virtualHosts().get()`, where the two aren't in sync. I don't think it is a good idea to have two AtomicReferences here. Instead, we should create a new class that is just a struct containing perRouteInterceptors and virtualHosts, and have a single AtomicReference for that class.

Reply:
+1